### PR TITLE
refactored preprocessor/merged mt feature extraction in encoder/fixed some bugs

### PR DESCRIPTION
**Pull request by LeungTsang**
preprocessor:

1.preprocessor is separated from dataset. dataset is no longer recursive. instead, dataset accepts a preprocessor object for data preprocessing
2.transformation can be defined in arbitary order. e.g. crop and band filter can be the first step to reduce computation.
tile is replaced by sliding inference in evalutor, which avoids loading images from disk multiple times and allows 3.generating complete prediction for a whole image.
4.enable verification of data size and dimension

encoder/decoder:
1.incorporate multi temporal forward in encoder/base.py  instead
2.squeeze temporal dimension safely

fix some bugs:
multi temporal output
save best model

TO DO:
check if everything works
merge segmentation and regression tasks as much as possible
fix bugs e.g. regevaluator

### CODE REVIEW

**Review by gle-bellier — COMMENTED** [review]
Hi, 
Thank you for the pull request, the preprocessing refactoring was needed and much appreciated. The same for the sliding inference of the evaluator. 
I included several comments for this review, most of them are related to comments about the shape of the tensors, but one of them requires special attention: we can not merge multi-temporal frames on the batch dimension otherwise we may face OOM.
I have not run any experiments with the codebase for now, but I will soon.

**Review by KerekesDavid — COMMENTED** [review]
I added some comments/questions with my thoughts as well, mostly about design decisions that were made deliberately in the previous version. (And might be worth considering)

**Comment by KerekesDavid** [comment]
> tile is replaced by sliding inference in evalutor, which avoids loading images from disk multiple times

I actually thought about this, and had a cache implementation for this in an iteration of the previous version. When I benchmarked it however it turns out the files are already memcached by the operating system (at least on my machines). So did not improve performance at all, and I left it out in the final pull request to reduce code complexity.

File: pangaea/datasets/ai4smallfarms.py [hunk 1]
```diff
@@ -9,7 +9,7 @@
 from tifffile import imread
 
 from pangaea.datasets.base import GeoFMDataset
-
+from pangaea.engine.data_preprocessor import BasePreprocessor
```
[the remark this answers is not in the document]
Comment (VMarsocci) [reply pangaea/datasets/ai4smallfarms.py:12]: Hi, on this I disagree :)

I see your point, but in deep learning community is very common to have the transformations (i.e. the preprocessing) in the dataloader. 

So I think it will be clear for people who read the code. And in general the code will be cleaner.

For similar reasons, I don't think people who want to add new datasets will forget, especially, if we document it in the readme.

[the remark this answers is not in the document]
Comment (KerekesDavid) [reply pangaea/datasets/ai4smallfarms.py:12]: In my 10 years of deep learning I've only ever seen it in repositories that were dealing with a single dataset, but it might be that I'm just old-school 😄

[the remark this answers is not in the document]
Comment (LeungTsang) [reply pangaea/datasets/ai4smallfarms.py:12]: I rethink about it and we can do a single dataset wrapper (not recursive) to combine a raw dataset and the transforms.

File: pangaea/run.py [hunk 2]
```diff
@@ -166,7 +161,7 @@ def main(cfg: DictConfig) -> None:
             num_workers=cfg.num_workers,
             pin_memory=True,
             # persistent_workers=True causes memory leak
-            persistent_workers=False,
+            persistent_workers=True,
```
[the remark this answers is not in the document]
Comment (LeungTsang) [reply pangaea/run.py:165]: ok

File: pangaea/encoders/base.py [hunk 3]
```diff
@@ -134,29 +118,56 @@ def freeze(self) -> None:
...
         Returns:
-            list[torch.Tensor]: list of the embeddings for each modality. For single-temporal encoders, the list's
-            elements are of shape (B, embed_dim, H', W'). For multi-temporal encoders, the list's elements are of shape
-            (B, C', T, H', W') with T the number of time steps if the encoder does not have any time-merging strategy,
-            else (B, C', H', W') if the encoder has a time-merging strategy (where C'==self.output_dim).
+            torch.Tensor: embedding generated by the encoder.
         """
         raise NotImplementedError
 
+    def forward(self, image: dict[str, torch.Tensor]) -> list[torch.Tensor]:
```
[the remark this answers is not in the document]
Comment (LeungTsang) [reply pangaea/encoders/base.py:135]: ok

File: pangaea/encoders/base.py [hunk 4]
```diff
@@ -134,29 +118,56 @@ def freeze(self) -> None:
...
+        else:
+            if t == 1:
+                return self.native_forward(image)
+            else:
+                return self.naive_multi_temporal_forward(image)
+
+
+    def naive_multi_temporal_forward(self, image: dict[str, torch.Tensor]) -> list[torch.Tensor]:
+        b, c, t, h, w = image[list(image.keys())[0]].shape
+        image = {k: v.transpose(1, 2).contiguous().view(-1, c, 1, h, w) for k, v in image.items()}
```
[the remark this answers is not in the document]
Comment (LeungTsang) [reply pangaea/encoders/base.py:148]: ok, I will make it optional

File: pangaea/run.py [hunk 5]
```diff
@@ -56,6 +57,8 @@ def main(cfg: DictConfig) -> None:
     fix_seed(cfg.seed)
     # distributed training variables
     rank = int(os.environ["RANK"])
+    cfg.task.trainer.use_wandb = cfg.task.trainer.use_wandb and rank == 0
```
[the remark this answers is not in the document]
Comment (LeungTsang) [reply pangaea/run.py:60]: ok

File: pangaea/decoders/base.py [hunk 6]
```diff
@@ -1,7 +1,10 @@
+import math
+import torch
 import torch.nn as nn
+import torch.nn.functional as F
 
 from pangaea.encoders.base import Encoder
-
+import matplotlib.pyplot as plt
```
[the remark this answers is not in the document]
Comment (LeungTsang) [reply pangaea/decoders/base.py:7]: ok

File: pangaea/engine/data_preprocessor.py [hunk 7]
```diff
@@ -1,386 +1,160 @@
...
 
+from typing import Callable, Dict, List, Optional, Sequence, Union, Tuple
+import copy
 
-class RichDataset(Dataset):
-    """Dataset wrapper to add preprocessing steps."""
 
-    def __init__(self, dataset: GeoFMDataset, encoder: Encoder):
-        """Initialize the RichDataset.
+def build_preprocessor(preprocessing_cfg, dataset_cfg, encoder_cfg):
```
[the remark this answers is not in the document]
Comment (LeungTsang) [reply pangaea/engine/data_preprocessor.py:18]: ok

File: pangaea/engine/trainer.py [hunk 8]
```diff
@@ -11,7 +12,7 @@
 from torch.optim.optimizer import Optimizer
 from torch.utils.data import DataLoader
 from pangaea.utils.logger import RunningAverageMeter, sec_to_hm
-
+import wandb
```
[the remark this answers is not in the document]
Comment (LeungTsang) [reply pangaea/engine/trainer.py:15]: seems wandb is in the requirements.txt anyway. I think just installing it won't bother that much for people don't want to use it. If installing wandb is indeed a pain, we still need to rewrite the way to optionally import wandb as the previous one is weird

**Comment by LeungTsang** [comment]
> Did you see any performance improvements in your case?
i know that the files will be in memory after the first epoch. however, the images files still need to be decoded and preprocessed. the old tiling bothers me a lot as it decodes the full, large images from 25 to 64 times to finish an inference (224/128 input size, 1024 image size), which makes no senses and cost huge overhead.  meanwhile reading an intact image means you can easily visualize the corresponding prediction / perform test-time augmentation and other features to be implemented.

**Comment by KerekesDavid** [comment]
The visualization can be easily solved by having an ID tag in the output dictionary so we can reconstruct the images.

> however, the images files still need to be decoded and preprocessed. the old tiling bothers me a lot as it decodes the full, large images from 25 to 64 times to finish an inference (224/128 input size, 1024 image size), which makes no senses and cost huge overhead. 

This is what I thought as well, but by tiling first you don't do a lot of unnecessary preprocessing, and it seemed like in the end I did not gain any performance by caching the preprocessed images and tiling at the end of the pipeline.

But if you can show me actual performance improvements, I'm all in for this change. 👍

**Comment by LeungTsang** [comment]
**1. The preprocessor is again redesigned.** 

1.1 Previous GeoFMDataset is renamed to RawFMDataset and current GeoFMDataset is a wrapper class to combine RawFMDataset and Preprocessor. 

1.2 The Preprocessor is initialized by preprocessor_cfg, dataset_cfg, encoder_cfg. a list of defined preprocessor will then be initialized with data statistics/info (e.g., data mean/std) being tracked to ensure preprocessors works properly in any order. 

1.3 BandAdaptor is split into BandFilter and BandPadding. So BnadFilter in the beginning and BandPadding in the end to avoid operating trivial bands.

1.4 Tile is replaced by sliding_inference in evaluator. It avoids load, decode, preprocess images multiple times when evaluating and its implementation is more straightforward. Getting intact images also facilitates potential operations during inference.

1.5 Add FocusRandomCrop for sparse annotated data. For example MADOS has its most label ignored and RandomCrop can hardly crop regions with valid labels. FocusRandomCrop guarantee crops have valid labels.

1.6 Remove ImportanceRandomCrop at the moment. I feel ImportanceRandomCrop will not work as expected. The class imbalance is mainly due to small classes only appear in few images. It may not be helpful to crop the most important region in single image and the resulting data distribution is undetermined. I also don't think we think we should handle data imbalance ourselves but the foundation models should.

1.7 Add RandomResizedCrop.

**Comment by LeungTsang** [comment]
**2. The encoders are partially refactored and bugs are fixed.**

2.1 gfmswin, satlasnet, spectralgpt do not resume the 2D feature layout correctly. They are fixed now.

2.2 For models whose output is already multi-scale pyramid features (swin), the Feature2Pyramid neck in upernet is skiped. 

2.3 An default function (naive_reshape_to_2d) to reshape transformer output to 2D layout is used when possible, to make the code cleaner.

2.4 Single/multi temporal model vs single/multi temporal input are dealt with in the forward method of parent class encoder instead of in decoder. The native forward for each encoder is renamed to simple_forward. The code can be shared in this way. 

2.5 The naive method (naive_multi_temporal_forward) for single temporal model to perform multi temporal forward is added. Loop mode iterates over T dimension to extract features one by one and batch mode merge T dimension into B to extract features in one forward.

2.6 SiamUperNet feed single frame to encoder but the encoder is configured to accept 2 frames. An enforce_single_temporal method is added to turn the encoder to single-temporal setting.

2.7 I try to redesign the encoder configs to make all the above stuff works. They are not satisfied in the end so I stop commenting them. I try to classify arguments into base encoder arguments (essential properties of the encoder) and model-specific arguments and merge those with the same meaning . However, it is cumbersome and there are always either special casesas the encoder implementations are heterogeneous. 

2.8 Now my idea is to put the base encoder arguments into an independent dict accepted by the base class. Users will have to fill these properties of the encoder strictly according to the format in config to ensure other module works with the encoder.  The free-form model-specific arguments are passed to the actual init function of the encoder. Separating the two groups of arguments minimizes the work to add new encoders.

example config:

```
###_model-specific arguments_###
img_size: 224
embed_dim: 768
...

###_base encoder arguments_###
meta_base:
    model_name: good
    download_url: xxx
    multi_temporal: 2
    input_size: 224 (so no conflicts with the img_size) 

class Encoder():
    def __init__(self, model_name, input_size, ...):
         self.input_size = input_size
         self.model_name = model_name
         ...

class GoodEncoder(Encoder):
    def __init__(self, img_size, embed_dim, ...., meta):
         super().__init__(**meta)
         self.img_size = img_size
         self.block = Block(embed_dim)
         ....
```

**Comment by LeungTsang** [comment]
**3. Other changes**

3.1 Computing overall mean MSE in Regression evaluator is still incorrect because batch size is not even during testing and the metric is not reduced across all GPUs. It is fixed now.

3.2 The current main does not save best checkpoint correctly. Best checkpoint only stores the reference to the model weight but not the actual current values. A quick fix is to make a deepcopy but I prefer to immediately save it to disk to avoid the memory cost.

3.3 Skipping NaN loss is replaced by raising an error. The initial motivation was the same as 1.5 but the model was still not properly trained and it might suppress other issues.

3.4 The parameters for the parent class (Dataset/Encoder) is moved to **kwargs to avoid assigning dozens of parameters repeatedly.

3.5 In trainer, num_classes is obtained from decoder instead of dataset, which is easier.

3.6 Correct the left time compute.

**Comment by KerekesDavid** [comment]
First of all that's a lot of things Liang, thanks for all the work!

Just some initial thoughts, below I don't have enough time to review this today, so I'lll come back to it on Monday in detail.

- This PR is an a bit too unfocused, there are ~3000 lines changed, and I think a lot of the stuff could be broken up into smaller PR-s that are easier to reason about. Eg. most of the fixes in the "3 .Other Changes" section could just be merged independently of the work here.
- Merge from main so there are no merge conflicts before the reviews.
- The change detection stuff that @SebastianGer  and @SebastianHafner is working on relies on importance cropping. The focus crop might be able to replace this, but they should take a look at it.
